### ISSUE TITLE
Support for Map

### ISSUE THREAD

**Issue by 13h3r** [pureconfig/pureconfig#6]
Add support of Map. Here it is an example:

Config:

``` hocon
params {
  a = 1
  b = 2
  c = 3
}
```

should maps into

``` scala
case class Root(params: Map[String, Int])
```

**Comment by melrief** [comment]
Nice idea. One question that I have is what kind of flexibility this system should have: should the Map contain an entire branch (from the branch point to the leafs) of the configuration or support only one level? In other words

``` scala
params.a=1
params.b.foo=2
params.c.bar.d=3
```

should return `Root(Map("a" -> 1, "b.foo" -> 2, "c.bar.d" -> 3))` or error because it expects "b" and "c" to be `Int`? The first case is very flexible because it allows to have "dynamic" parts of the configuration that are not known at compile time but can be used at runtime. This would be my choice but I wonder what kind of implications this would have with Map whose values are classes. The second approach is easier to implement but way less flexible.

What do you think should be the behaviour?

**Comment by 13h3r** [comment]
As for me we should choice second approach. If I need to work with config in dynamic style then mapping them into static object probably not the best way.

Lets take a look to your example. What will we have after parsing? `Map[String, Any]`? This API is similar to original API and unsafe. Describing config in terms of case classes is the way to type safety.

**Comment by melrief** [comment]
Implemented with d4ecdb4. I'm closing this, I will reopen it if you think there are problems with the implementation. As side note, I'm considering adding also the first case with a special type wrapper. But first I have to see if there is a proper way to do it.

**Comment by mareksabo** [comment]
> If I need to work with config in dynamic style then mapping them into static object probably not the best way.

What if you need to pass that entire configuration part to another class/method? Then you want the flexibility to change anything without breaking any classes. I wanted to pass one part of the configuration to `SparkConf` and I came up with this solution:

```scala
case class SparkPropertiesConfig(spark: com.typesafe.config.Config)

// implicit method for importing spark properties SparkSession.builder.configureSparkProperties(config.spark)
implicit class RichSparkConf(val conf: SparkSession.Builder) {
  def configureSparkProperties(config: SparkPropertiesConfig): SparkSession.Builder = {
    config.spark.entrySet().foreach { entry => conf.config(entry.getKey, entry.getValue.unwrapped().toString) }
    conf
  }
}
```

I hope it will be useful.
